File: walking_robot_sim/src/quadropted_controller_cpp/src/controllers/pid_controller.cpp
#include "quadropted_controller_cpp/controllers/pid_controller.hpp"

#include <cmath>

namespace quadropted {

PIDController::PIDController(double kp, double ki, double kd) : kp_(kp), ki_(ki), kd_(kd) {}

std::array<double, 2> PIDController::run(double roll, double pitch, double current_time) {
    std::array<double, 2> error = {desired_roll_pitch_[0] - roll, desired_roll_pitch_[1] - pitch};

    if (last_time_ < 0.0) {
        last_time_ = current_time;
        return {0.0, 0.0};
    }

    double step = current_time - last_time_;
    if (step < 1e-6) {
        return {0.0, 0.0};
    }

    for (int i = 0; i < 2; ++i) {
        i_term_[i] += error[i] * step;
        if (i_term_[i] < -max_i_)
            i_term_[i] = -max_i_;
        else if (i_term_[i] > max_i_)
            i_term_[i] = max_i_;

        d_term_[i] = (error[i] - last_error_[i]) / step;
    }

    last_time_ = current_time;
    last_error_ = error;

    std::array<double, 2> result;
    for (int i = 0; i < 2; ++i) {
        result[i] = kp_ * error[i] + ki_ * i_term_[i] + kd_ * d_term_[i];
    }

    return result;
}

void PIDController::reset(double current_time) {
    last_time_ = current_time;
    i_term_ = {0.0, 0.0};
    d_term_ = {0.0, 0.0};
    last_error_ = {0.0, 0.0};
}

void PIDController::set_desired(double roll, double pitch) {
    desired_roll_pitch_[0] = roll;
    desired_roll_pitch_[1] = pitch;
}

}  // namespace quadropted

File: walking_robot_sim/src/tests/old/RobotController/PIDController.py
#!/usr/bin/env python3
# Author: lnotspotl, abutalipovvv
import rclpy
from rclpy.time import Time
import numpy as np

class PID_controller:
    def __init__(self, kp, ki, kd):
        self.kp = kp
        self.ki = ki
        self.kd = kd

        self.desired_roll_pitch = np.array([0.0, 0.0])

        self.I_term = np.array([0.0, 0.0])
        self.D_term = np.array([0.0, 0.0])

        self.max_I = 0.2
        self.last_error = np.array([0.0, 0.0])
        self.last_time = None

    def run(self, roll, pitch):
        error = self.desired_roll_pitch - np.array([roll, pitch])

        t_now = self.get_time_in_seconds()
        if self.last_time is None:
            self.last_time = t_now
            return np.array([0.0, 0.0])  

        step = t_now - self.last_time

        if step < 1e-6:  
            return np.array([0.0, 0.0])

        self.I_term += error * step

        for i in range(2):
            if self.I_term[i] < -self.max_I:
                self.I_term[i] = -self.max_I
            elif self.I_term[i] > self.max_I:
                self.I_term[i] = self.max_I

        self.D_term = (error - self.last_error) / step

        self.last_time = t_now
        self.last_error = error

        P_ret = self.kp * error
        I_ret = self.I_term * self.ki
        D_ret = self.D_term * self.kd

        return P_ret + I_ret + D_ret

    def reset(self):
        self.last_time = self.get_time_in_seconds()
        self.I_term = np.array([0.0, 0.0])
        self.D_term = np.array([0.0, 0.0])
        self.last_error = np.array([0.0, 0.0])

    def set_desired_RP_angles(self, des_roll, des_pitch):
        self.desired_roll_pitch = np.array([des_roll, des_pitch])

    def get_time_in_seconds(self):
        return Time().seconds_nanoseconds()[0] + Time().seconds_nanoseconds()[1] * 1e-9
